mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,332 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict, Tuple
+
+import jax
+import jax.numpy as jnp
+from chex import Array, PRNGKey
+from gigastep.gigastep_env import GigastepEnv
+from jumanji import specs
+from jumanji.types import StepType, TimeStep, restart
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation, ObservationGlobalState, State
+
+if TYPE_CHECKING:  # https://github.com/python/mypy/issues/6239
+    from dataclasses import dataclass
+else:
+    from flax.struct import dataclass
+
+
+@dataclass
+class GigasteplState:
+    """Wrapper around a Gigastep state to provide necessary attributes for jumanji environments."""
+
+    state: State
+    key: PRNGKey
+    step: int
+    Adverseary_action: Array
+
+
+class GigastepWrapper(Wrapper):
+    """
+    Wraps a GigaStep environment so that its API is compatible with Jumanji environments.
+
+    Args:
+        env: The GigaStep environment to be wrapped.
+        time_limit (int): The maximum duration of each episode, in seconds. Defaults to 500.
+        has_global_state (bool): Whether the environment has a global state. Defaults to False.
+        get_adversary_obs (bool): Whether to retrieve adversary observations for the global state.
+        Defaults to False.
+        get_adversary_actions (bool): Whether to retrieve adversary actions for the global state.
+        Defaults to False.
+
+    Methods:
+        reset: Resets the environment and returns a GigasteplState and the restart.
+        step: Takes an action and returns the next GigasteplState and timestep.
+        action_mask: Returns the action mask for each agent.
+        get_global_state: Returns the global state of the environment.
+        observation_spec: Returns the observation spec for the environment.
+        action_spec: Returns the action spec for the environment.
+        reward_spec: Returns the reward spec for the environment.
+        discount_spec: Returns the discount spec for the environment.
+        _split_obs_and_state: Splits the observation and state into separate arrays for each team.
+        get_wining_team: Returns the winning team based on the number of alive agents.
+        adversary_policy: Chooses an action for the adversary.
+    """
+
+    def __init__(
+        self,
+        env: GigastepEnv,
+        time_limit: int = 500,
+        has_global_state: bool = False,
+        get_adversary_obs: bool = False,
+        get_adversary_actions: bool = False,
+    ):
+
```

Comment:
We put this under `__init__` and put the method docs inside the functions :+1: 

```suggestion
    def __init__(
        self,
        env: GigastepEnv,
        time_limit: int = 500,
        has_global_state: bool = False,
        get_adversary_obs: bool = False,
        get_adversary_actions: bool = False,
    ):
    """
    Wraps a GigaStep environment so that its API is compatible with Jumanji environments.

    Args:
        env: The GigaStep environment to be wrapped.
        time_limit (int): The maximum duration of each episode, in seconds. Defaults to 500.
        has_global_state (bool): Whether the environment has a global state. Defaults to False.
        get_adversary_obs (bool): Whether to retrieve adversary observations for the global state.
        Defaults to False.
        get_adversary_actions (bool): Whether to retrieve adversary actions for the global state.
        Defaults to False.
    """
```